### DISCUSSION TITLE
lxqt-panel: dev: how to test changes in real time by replacing own panel ?

### DISCUSSION THREAD

**Discussion by AdelKS** [lxqt/lxqt#2241] [Q&A] [answered]
Hello! 

I tried to rebase my [PR on lxqt-panel](https://github.com/lxqt/lxqt-panel/pull/1698) against recent changes and it successfully rebased by the panel crashes when I start typing to search.

Is there a way to replace my current panel with the one I am modifying in Qt Creator to quickly test, debug, and iterate ?

PS: I could not find quickly enough a dev wiki on how to work on LXQt

**Comment by yan12125** [comment]
Stop lxqt-panel and run the modified verison? Usually I use lxqt-config-session to stop lxqt-panel, and a process manager like htop should work work. To run custom lxqt-panel, I type the command from a terminal. Running it from Qt Creator should also work I guess.

**Reply by AdelKS** [reply]
Okay I stopped it from `lxqt-config-session`. But running from Qt Creator gives me a black bar :thinking:

**Reply by AdelKS** [reply]
It gives me this when I run it from the terminal
```shell
> $ ./lxqt-panel                                                                                                                                    
Failed to create wl_display (No such file or directory)
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
"Plugin \"mainmenu\" not found."
"Plugin \"desktopswitch\" not found."
"Plugin \"quicklaunch\" not found."
"Plugin \"taskbar\" not found."
"Plugin \"statusnotifier\" not found."
"Plugin \"kbindicator\" not found."
"Plugin \"worldclock\" not found."
WinIdChange 600006 handle QWidgetWindow(0x558292b698b0, name="LXQtPanel panel1Window") QScreen(0x558292a8a9a0, name="DisplayPort-0")
```
so I suppose it's trying to load with Wayland ?

**Reply by AdelKS** [reply]
No it's not Wayland: I tried with `-platform xcb` to silence those warnings and still black bar. It's because of the other plugins that are not found.

**Reply by AdelKS** [reply]
I cannot figure it out, I will await your help so I can proceed further, thanks!

**Reply by yan12125** [reply]
Oops, I forgot lxqt-panel cannot find plugins out-of-box. You need to run `sudo make install` first. The command may be slightly different for Qt Creator.

**Reply by palinek** [reply]
Directories for searching the plugins can be driven by env. variables, e.g.:
```
make -j4 && LXQT_PANEL_PLUGINS_DIR="$(echo plugin-* | tr ' ' ':')" LXQTPANEL_PLUGIN_PATH="$(echo plugin-* | tr ' ' ':')" panel/lxqt-panel -c ../test_panel.conf
```

**Reply by yan12125** [reply]
Nice! Didn't know this trick.

**Reply by AdelKS** [reply]
Thanks @palinek it works! And I can now debug within Qt Creator. The issue I have now is that when the panel segfaults the entire desktop becomes unresponsive, any idea how I can mitigate this ?

**Reply by AdelKS** [reply]
Actually apps are responsive, it's just that the keyboard and mouse remain grabbed within the stalling lxqt-panel, since it is on a break point :thinking:

**Reply by yan12125** [reply]
> The issue I have now is that when the panel segfaults

There were reports that some plugins crash under Wayland (https://github.com/lxqt/lxqt-panel/pull/1124). Probably XWayland is worth a try.

> since it is on a break point

You can run a debugger remotely (ex: gdb via ssh).

**Reply by AdelKS** [reply, marked as answer]
Okay, I found a complex way to do it, but it works: use another TTY to tell X11 to ungrab mouse and keyboard, found how to do that from [here](https://unix.stackexchange.com/questions/40458/command-for-forcing-a-pointer-ungrab-captured-mouse-release)

Summary on how to debug lxqt-panel with Qt Creator

1. Setup runtime environment with the variables `LXQT_PANEL_PLUGINS_DIR` and `LXQTPANEL_PLUGIN_PATH` (`echo plugin-* | tr ' ' ':')` is run manually and its output is set into those env vars)
2. Tell x11 to accept the special command that tells it to ungrab mouse and keyboard `setxkbmap -option grab:break_actions`
3. Stop the LXQt's default panel (the one provided by the distro), this can be done with the program `lxqt-config-session`
4. Run lxqt-panel on Qt Creator on debug mode and with the wanted break points.
5. Once a breakpoint is triggered, switch to another TTY, run `export DISPLAY=':0'`, then run `xdotool key XF86Ungrab`. Go back to the LXQt session TTY, now the keyboard and mouse are un-grabed and can be used for GUI debugguing within Qt Creator (see the stack, the value of the variables, do step by step instructions, check previous stacks... etc)

> You can run a debugger remotely (ex: gdb via ssh).

I find command line debugging very inconvenient.

> There were reports that some plugins crash under Wayland (https://github.com/lxqt/lxqt-panel/pull/1124). Probably XWayland is worth a try.

I am using LXQt on x11. Thanks @yan12125. I could fix the issues I have been having when I could debug properly.

**Reply by AdelKS** [reply]
@yan12125 @palinek would you have some time to have a look at [my PR](https://github.com/lxqt/lxqt-panel/pull/1698), I took the time to rebase and fix the issues.

**Reply by yan12125** [reply]
It's now in my very long PR review queue :)

**Comment by tsujan** [comment]
You may not like what I'm going to say but, IMO, Qt Creator is rather an annoying obstacle. A terminal emulator and a lightweight code editor is all one may need for coding and debugging.

**Reply by AdelKS** [reply]
It's not like I am not going to like it. Tastes and colors are not universal. And of course one can even code in something like `gedit`, heck, even `nano`. However, I think that having a graphical interface for debugging saves a lot of time. You probably haven't tried Qt Creator's debugging features ? `vscode` can do the same but I haven't had enough luck with it, and Qt Creator is just plainly better for C++, even though it's way less flexible.

Also, would you have time to have a look at [my PR](https://github.com/lxqt/lxqt-panel/pull/1698) ? :smile:
